File: oc/UIPickerViewDemo/UIPickerViewDemo/FlagViewController.m
//
//  FlagViewController.m
//  UIPickerViewDemo
//
//  Created by dfang on 2018-1-6.
//  Copyright © 2018年 east. All rights reserved.
//

#import "FlagViewController.h"
#import "FlagView.h"

@interface FlagViewController ()

@property (nonatomic, strong) UIPickerView *pickerview;
@property (nonatomic, strong) NSArray *flags;

@end

@implementation FlagViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.view.backgroundColor = [UIColor whiteColor];
    
    self.pickerview.delegate = self;
    self.pickerview.dataSource = self;
    [self.view addSubview:self.pickerview];
    
}

- (UIPickerView *)pickerview {
    if (_pickerview == nil) {
        _pickerview = [[UIPickerView alloc] initWithFrame:self.view.frame];
    }
    
    return _pickerview;
}

- (NSArray *)flags {
    if (_flags == nil) {
        _flags = [NSArray arrayWithContentsOfFile: [[NSBundle mainBundle] pathForResource:@"flags" ofType:@"plist"]];
    }
    
    return _flags;
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (NSInteger)numberOfComponentsInPickerView:(nonnull UIPickerView *)pickerView {
    return 1;
}

- (NSInteger)pickerView:(nonnull UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    return self.flags.count;
}

- (UIView *)pickerView:(UIPickerView *)pickerView viewForRow:(NSInteger)row forComponent:(NSInteger)component reusingView:(UIView *)view {
    FlagView *flagView = (FlagView *)view;
    if (flagView == nil) {
        flagView = [[FlagView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, 100)];
    }
    
    NSDictionary *item = self.flags[row];
    flagView.nameLabel.text = [item valueForKey:@"name"];
    flagView.iconImageView.image = [UIImage imageNamed:[item valueForKey:@"icon"]];
    
    return flagView;
}

- (CGFloat)pickerView:(UIPickerView *)pickerView rowHeightForComponent:(NSInteger)component {
    return 100;
}

@end

File: swift/UIPickerViewDemo/UIPickerViewDemo/FlagViewController.swift
//
//  FlagViewController.swift
//  UIPickerViewDemo
//
//  Created by dfang on 2018-1-5.
//  Copyright © 2018年 east. All rights reserved.
//

import UIKit

class FlagViewController: ViewController, UIPickerViewDataSource, UIPickerViewDelegate {
    
    let flags = NSArray.init(contentsOfFile: Bundle.main.path(forResource: "flags", ofType: "plist")!)!
    let pickerView = UIPickerView.init()
    
    override func viewDidLoad() {
        pickerView.frame = self.view.frame
        pickerView.dataSource = self
        pickerView.delegate = self
        
        self.view.addSubview(pickerView)
    }
    
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return flags.count
    }
    
    func pickerView(_ pickerView: UIPickerView, viewForRow row: Int, forComponent component: Int, reusing view: UIView?) -> UIView {
        var flagView = view as? FlagView
        if (nil == flagView) {
            flagView = FlagView.init(frame: CGRect.init(x: 0, y: 0, width: 100, height: 100))
        }
        
        let dic = flags[row] as! NSDictionary
        flagView?.nameLabel.text = dic.value(forKey: "name") as? String
        flagView?.iconImageView.image = UIImage.init(named: dic.value(forKey: "icon") as! String)
        
        return flagView!
    }
    
    func pickerView(_ pickerView: UIPickerView, rowHeightForComponent component: Int) -> CGFloat {
        return 100
    }
}
